digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java

```diff
@@ -341,18 +347,32 @@ CommandStatus createNextCommandStatus(
             }
             while (Instant.now().getEpochSecond() < timeout);
 
+            final ObjectNode nextStatus = previousStatus.deepCopy();
+            nextStatus.set("executor_state", previousExecutorStatus);
+            final Optional<String> errorMessage = getErrorMessageFromTask(cluster, taskArn, client);
+
             final String outputArchivePathName = "archive-output.tar.gz";
             final String outputArchiveKey = createStorageKey(commandContext.getTaskRequest(), outputArchivePathName); // url format
             // Download output config archive
             final TemporalProjectArchiveStorage temporalStorage = createTemporalProjectArchiveStorage(commandContext.getTaskRequest().getConfig());
             try (final InputStream in = temporalStorage.getContentInputStream(outputArchiveKey)) {
                 ProjectArchives.extractTarArchive(commandContext.getLocalProjectPath(), in); // IOException
+            } catch (StorageFileNotFoundException ex) {
+                if (statusCode == 0) {
+                    String loggingFinishedAt = previousExecutorStatus.has("logging_finished_at") ? previousExecutorStatus.get("logging_finished_at").asText() : null;
+                    // The python script itself finished successful but the following command (e.g. tar zcf or curl -X PUT) failed ?
```

Comment:
How about outputting the results of the `tar zcf` and `curl -X PUT` like https://github.com/treasure-data/digdag/blob/a5fbe9160fc254898d864dee54a5e96fdea7ce3e/digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java#L751 ?

Reply:
Actually tar stderr is already captured such as `tar: .digdag/tmp: file changed as we read it`. tar stdout is no need to be logged.

Curl's output (stdout/stderr) is suppressed with silent mode because Muga requested not to leak S3 pre-signed URL at that time.
However, it's not shown to customers in TD production. 

I'm considering to replace `--silent/-s` with `--show-error` to capture errors.

```
 -S, --show-error    Show error. With -s, make curl show errors when they occur
 -s, --silent        Silent mode (don't output anything)
```

Reply:
How do you think? Take risks for leaking presigned S3 url or keep as it is.

Reply:
The intentional silent mode seems reasonable to me. How about just showing the exit code of curl like this?

```java
         bashArguments.add(s("curl --retry %d --retry-connrefused%s -s -X PUT -T %s -L \"%s\"", retryUploads, curlFailOptOnUploads ? " --fail" : "", outputProjectArchivePathName, outputProjectArchiveDirectUploadUrl));
+        bashArguments.add(s("echo \"curl_exit_code: $?\"");
         bashArguments.add(s("echo \"%s\"", ECS_END_OF_TASK_LOG_MARK));
```

Reply:
Hmm... `curl`'s exit code is 0 whenever it receives HTTP response including 4xx or 5xx...
```
$ curl https://localhost/hello/world
curl: (7) Failed to connect to localhost port 443: Connection refused
$ echo $?
7
$ curl https://192.0.2.0/hello/world
curl: (7) Failed to connect to 192.0.2.0 port 443: Operation timed out
$ echo $?
7
$ curl https://github.com/invalidpath/hello/world
Not Found
$ echo $?
0
```

Reply:
`-f` looks to work.

```
$ curl -f https://github.com/invalidpath/hello/world
curl: (22) The requested URL returned error: 404
$ echo $?
22
```

Reply:
@komamitsu 

http status of curl command is already shown in the stdout and thus we don't need to explicitly echo curl status code.
try

```
# without silent option just for debugging
$ curl --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80
Warning: Transient problem: HTTP error Will retry in 1 seconds. 4 retries 
Warning: left.
Warning: Transient problem: HTTP error Will retry in 2 seconds. 3 retries 
Warning: left.
Warning: Transient problem: HTTP error Will retry in 4 seconds. 2 retries 
Warning: left.
{"status":500,"message":"Internal Server Error"}{"status":503,"message":"Service Unavailable"}{"status":503,"message":"Service Unavailable"}{"status":200,"message":"OK"}MBP13-myui:~ myui$ curl --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80 
{"status":200,"message":"OK"}

# same as the current digdag code with a silent option
$ curl --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80
$ curl --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80 -s 2>/dev/null

{"status":500,"message":"Internal Server Error"}{"status":503,"message":"Service Unavailable"}{"status":503,"message":"Service Unavailable"}{"status":504,"message":"Gateway Timeout"}{"status":500,"message":"Internal Server Error"}
```

Note `--retry` option is introduced to avoid temporary http errors. See https://curl.se/docs/manpage.html

BTW, we already introduced `--fail` (-f in short) option
https://github.com/treasure-data/digdag/blame/a5fbe9160fc254898d864dee54a5e96fdea7ce3e/digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java#L750

However, it should not be enabled by the default because `it's intended to curl exit 0 on http errors since it's only for LogWatch logging` as I commented [here](https://github.com/treasure-data/digdag/blame/a5fbe9160fc254898d864dee54a5e96fdea7ce3e/digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java#L749).

We enable it for [the fatal command](https://github.com/treasure-data/digdag/blame/a5fbe9160fc254898d864dee54a5e96fdea7ce3e/digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java#L734) though.

Reply:
to use http status code, `-o /dev/null -w '%{http_code}\n'` can be used but it has less debugging info than the current options.

```
CURL_HTTP_STATUS_CODE=$(curl  --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80 -o /dev/null -w '%{http_code}\n' -s)
echo "CURL_HTTP_STATUS_CODE=$CURL_HTTP_STATUS_CODE"
```

> CURL_HTTP_STATUS_CODE=503

Reply:
> http status of curl command is already shown in the stdout and thus we don't need to explicitly echo curl status code. stdout is logged to cloudwatch logs.

Ah, you're right. So, `-f` option isn't needed.

My concern is not only about HTTP error but also connection error that silently finishes with non-zero exit code. Is there any curl option to show this kind of connection error?

I don't insist on the idea of `echo`ing the curl's exit code. Anything is ok as far as we can isolate the root cause of `io.digdag.spi.StorageFileNotFoundException: S3 file not found` 😃

Reply:
> Is there any curl option to show this kind of connection error?

`--show-error` option and relax `--silent` condition.

However, we are already getting at least `{"status":500,"message":"Internal Server Error"}{"status":503,"message":"Service Unavailable"}{"status":503,"message":"Service Unavailable"}{"status":504,"message":"Gateway Timeout"}{"status":500,"message":"Internal Server Error"}` in logs.

```
curl --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80
{"status":504,"message":"Gateway Timeout"}{"status":500,"message":"Internal Server Error"}{"status":500,"message":"Internal Server Error"}{"status":503,"message":"Service Unavailable"}{"status":500,"message":"Internal Server Error"}

curl --retry 4 --retry-connrefused https://instability.vercel.app/?errorRate=80 --show-error
Warning: Transient problem: HTTP error Will retry in 1 seconds. 4 retries 
Warning: left.
Warning: Transient problem: HTTP error Will retry in 2 seconds. 3 retries 
Warning: left.
Warning: Transient problem: HTTP error Will retry in 4 seconds. 2 retries 
Warning: left.
Warning: Transient problem: HTTP error Will retry in 8 seconds. 1 retries 
Warning: left.
{"status":504,"message":"Gateway Timeout"}{"status":500,"message":"Internal Server Error"}{"status":500,"message":"Internal Server Error"}{"status":503,"message":"Service Unavailable"}{"status":500,"message":"Internal Server Error"}
```

Reply:
```
When curl exits with a non-zero code, it will also output an error message (unless --silent is used). That error message may add some additional information or circumstances to the exit status number itself so the same error number can get different error messages
```

https://everything.curl.dev/usingcurl/returns#error-message

So, relaxing "--silent" with taking a risk for presigned S3 url.

Reply:
> Is there any curl option to show this kind of connection error?

Sorry, I remember I confirmed in https://github.com/treasure-data/digdag/pull/1603#discussion_r650610071 that connection errors are output to stderr.

According to your comment, we've already seen some log messages of the upload failure. That's apparently the cause of some `io.digdag.spi.StorageFileNotFoundException: S3 file not found`. Why don't you increase the retry count from 4 to more?

Reply:
> According to your comment, we've already seen some log messages of the upload failure.

true. Http return `{"status":504,"message":"Gateway Timeout"}` is already shown.

> Why don't you increase the retry count from 4 to more?

retry=4 is just an example for explanation.

We use retry=7 for upload by the default and it's already configurable through system config.

https://github.com/treasure-data/digdag/blob/v0_11/digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java#L77

At TD, we use

```
command_executor.ecs.retry_task_scripts_downloads: 11
command_executor.ecs.retry_task_output_uploads: 10
```

It's enough.

Reply:
Sounds reasonable.

We've already confirmed current retry count is enough, so it means `curl` isn't the cause of `S3 file not found` and only `tar zcf` is a potential cause?

Reply:
BTW, `curl -s` option suppresses error message of connection issue.
```
$ curl -s https://localhost/hello/world
$ echo $?
7
```
Hmm... Maybe showing the exit code of `curl` would be helpful if we keep using `-s` option.

Reply:
@komamitsu fixed to print exit code for curl https://github.com/treasure-data/digdag/pull/1603/commits/7869d01fb1d6618021fe54ad71262c2570a2e692 

Could you review it again? Also, force pushed to resolve conflicts with v0_11.